### PR TITLE
executor: stop fetch handles after closing.

### PR DESCRIPTION
**Pull request by hanfei1991**
terminate the index read as soon as possible.

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -414,6 +414,7 @@ func (e *XSelectIndexExec) Close() error {
 
 	e.taskCurr = nil
 	if e.taskChan != nil {
+		e.ctx.Cancel()
```
Comment (shenli) [inline executor/distsql.go]: Why not move Cancel() before the if statement?

**Comment by shenli** [comment]
@hanfei1991 CI failed.

File: executor/distsql.go (hunk 1 above)
Comment (hanfei1991) [inline executor/distsql.go]: Only double read need this

**Comment by hanfei1991** [comment]
It's hard to write unit test. A simple test can prove this improvement:
mysql> show create table BT_1;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                             |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| BT_1  | CREATE TABLE `BT_1` (
  `pid` int(11) DEFAULT NULL,
  `cnt` int(11) DEFAULT NULL,
  `act` int(11) DEFAULT NULL,
  KEY `k1` (`cnt`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

Before change:

mysql> select * from BT_1 where cnt < 8 and pid < 10000000 limit 1;
+------+------+------+
| pid  | cnt  | act  |
+------+------+------+
| 6845 |    0 |    2 |
+------+------+------+
1 row in set (1.16 sec)

mysql> select * from BT_1 where cnt < 8 and pid < 10000000 limit 1;
+------+------+------+
| pid  | cnt  | act  |
+------+------+------+
| 6845 |    0 |    2 |
+------+------+------+
1 row in set (3 min 52.17 sec)

After change:

mysql>  select * from BT_1 where cnt < 8 and pid < 10000000 limit 1;
+------+------+------+
| pid  | cnt  | act  |
+------+------+------+
| 6845 |    0 |    2 |
+------+------+------+
1 row in set (1.10 sec)

mysql>  select * from BT_1 where cnt < 8 and pid < 10000000 limit 1;
+------+------+------+
| pid  | cnt  | act  |
+------+------+------+
| 6845 |    0 |    2 |
+------+------+------+
1 row in set (1.14 sec)

**Comment by shenli** [comment]
LGTM
@coocood @tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**